src/LuaColor.cpp

```diff
@@ -0,0 +1,213 @@
+	double g = luaL_checknumber(L, 2);
+	double b = luaL_checknumber(L, 3);
+	double a;
+	if (lua_isnumber(L, 4)) a = luaL_checknumber(L, 4);
+	else a = 255.0;
+	sanity_check(r, g, b, a);
+	LuaColor::PushToLua(L, Color(r, g, b, a));
+	LUA_DEBUG_END(L, 1);
+	return 1;
+}
+
+static int l_color_call(lua_State *L)
+{
+	LUA_DEBUG_START(L);
+	double r = luaL_checknumber(L, 2);
+	double g = luaL_checknumber(L, 3);
+	double b = luaL_checknumber(L, 4);
+	double a;
+	if (lua_isnumber(L, 5)) a = luaL_checknumber(L, 5);
+	else a = 255.0;
+	sanity_check(r, g, b, a);
+	Color c = Color(r, g, b, a);
+	LuaColor::PushToLua(L, Color(r, g, b, a));
+	LUA_DEBUG_END(L, 1);
+	return 1;
+}
+
+static int l_color_tostring(lua_State *L)
+{
+	const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	luaL_Buffer buf;
+	luaL_buffinit(L, &buf);
+	char *bufstr = luaL_prepbuffer(&buf);
+	int len = snprintf(bufstr, LUAL_BUFFERSIZE, "Color(%i, %i, %i, %i)", c->r, c->g, c->b, c->a);
+	assert(len < LUAL_BUFFERSIZE); // XXX should handle this condition more gracefully
+	luaL_addsize(&buf, len);
+	luaL_pushresult(&buf);
+	return 1;
+}
+
+static int l_color_add(lua_State *L)
+{
+	const Color4ub *a = LuaColor::CheckFromLua(L, 1);
+	const Color4ub *b = LuaColor::CheckFromLua(L, 2);
+	LuaColor::PushToLua(L, *a + *b);
+	return 1;
+}
+
+static int l_color_mul(lua_State *L)
+{
+	if (lua_isnumber(L, 1)) {
+		const double s = lua_tonumber(L, 1);
+		const Color4ub *c = LuaColor::CheckFromLua(L, 2);
+		LuaColor::PushToLua(L, *c * s);
+	} else if (lua_isnumber(L, 2)) {
+		const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+		const double s = lua_tonumber(L, 2);
+		LuaColor::PushToLua(L, *c * s);
+	} else {
+		return luaL_error(L, "Cannot multiply a Color time Color");
+	}
+	return 1;
+}
+
+static int l_color_new_index(lua_State *L)
+{
+	Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	if (lua_type(L, 2) == LUA_TSTRING) {
+		const char *attr = luaL_checkstring(L, 2);
+		if (!strcmp(attr, "r")) {
+			c->r = luaL_checknumber(L, 3);
+		} else if (!strcmp(attr, "g")) {
+			c->g = luaL_checknumber(L, 3);
+		} else if (!strcmp(attr, "b")) {
+			c->b = luaL_checknumber(L, 3);
+		} else if (!strcmp(attr, "a")) {
+			c->a = luaL_checknumber(L, 3);
+		} else {
+			luaL_error(L, "Index '%s' is not available: use 'r', 'g', 'b' or 'a'", attr);
+		}
+
+	} else {
+		luaL_error(L, "Expected Color, but type is '%s'", lua_typename(L, lua_type(L, 2)));
+	}
+	LuaColor::PushToLua(L, *c);
+	return 1;
+}
+
+static int l_color_index(lua_State *L)
+{
+	const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	if (lua_type(L, 2) == LUA_TSTRING) {
+		const char *attr = luaL_checkstring(L, 2);
+		if (!strcmp(attr, "r")) {
+			lua_pushnumber(L, c->r);
+			return 1;
+		} else if (!strcmp(attr, "g")) {
+			lua_pushnumber(L, c->g);
+			return 1;
+		} else if (!strcmp(attr, "b")) {
+			lua_pushnumber(L, c->b);
+			return 1;
+		} else if (!strcmp(attr, "a")) {
+			lua_pushnumber(L, c->a);
+			return 1;
+		}
+	}
+	lua_getmetatable(L, 1);
```

Comment:
There's a function called `luaL_getmetafield` that does basically the same thing in an optimized one-liner, though it only supports string keys. Given the specific use case, I think that's appropriate.

We might also want to support accessing values by index and/or swizzling color values. E.g. `col[0] = x, col.bgr = col2.rgb`

Reply:
> There's a function called luaL_getmetafield that does basically the same thing in an optimized one-liner, though it only supports string keys. Given the specific use case, I think that's appropriate.

Done;

> We might also want to support accessing values by index and/or swizzling color values

Really? ...there are use-cases for all these things? ...Sure I miss them at all, may I ask where these use cases are?

Reply:
>Done
You sure? It doesn't seem to have made it in.

As far as use cases go, I forgot to stress the _might_; we're not using them now, so it's something to consider for later.